### PR TITLE
Split into crates

### PR DESCRIPTION
**Pull request by Alex-Fischman**
This PR splits the project up into multiple crates, which in theory will improve compile times. While this is a large change, most of the changes that I had to make were mechanical, except for the following:

- I made all `mod`s in all `lib.rs` and `mod.rs` public. This causes lots of missing documentation warnings. I did this because a) the code does actually need to be documented, and b) I believe that all modules in a crate should be public, and the module should decide what is exposed. While the second point is debatable, I don't think the first point is, so I don't think that this change should be reverted until we find a way to get the clippy lints without it.
- I removed all of the existing instances of `pub use` in `mod/lib.rs`. Since all modules are now public, this just creates an extra path to the same code, so it's unnecessary.
- I added some `pub use`s to the `lib.rs` in `solver`, so that clients of `solver` would not have to also depend on `smtlib`.
- We split the code up into crates as follows: `fly` and `smtlib` define common functionality among other crates; `bounded`, `inference`, `solver`, and `verify` all implement the functionality that the user cares about; and `temporal-verifier` defines the command line interface. While most files are in the crate that corresponds to the folder that they were in before this commit, a few files (such as `timing.rs` and `concurrent.rs`) got moved due to being in the wrong place in the dependency ordering.
- Finally, there was one more folder, `term` that we weren't quite sure what to do with, as it defines common functionality but also depends on `fly`. For now, we've left it as a submodule of the `fly` crate.

### CODE REVIEW

**Comment by Alex-Fischman** [comment]
I've left unused lines commented out in the root `Cargo.toml`. Someone who added those lines should look at them and determine if we still need them or not.

**Comment by wilcoxjay** [comment]
This is great, thanks so much for taking this on @Alex-Fischman!

As a side note, I did some further investigation into build times, and a significant portion (~50% by my estimate) of the end to end build time is spent recompiling the qalpha branch of `App::exec`. I believe this is due to the heavy use of generic instantiation in that branch. In the future it would be great to move that code into the qalpha file so that it does not need to be recompiled every time we build the app.

**Comment by Alex-Fischman** [comment]
I'm currently working on getting CI to pass by adding documentation, but I haven't used the code that I'm trying to document, so someone should check my work when I'm finished.

**Comment by Alex-Fischman** [comment]
Whoops, CI could've passed without getting docs done. There's some issue with tests in proc.rs. I'll look into it tomorrow.

**Comment by tchajed** [comment]
Could you stop adding any documentation in this PR, and turn off `#[warn(missing_docs)]` for now? It makes the PR much harder to review.

File: fly/src/term/cnf.rs [hunk 1]
```diff
@@ -180,7 +180,7 @@ pub fn term_to_cnf_clauses(t: &Term) -> Vec<Term> {
 }
 
 impl Cnf {
-    #[allow(dead_code)]
+    /// Build a Term in CNF out of any Term
```
Comment (tchajed) [inline fly/src/term/cnf.rs:183]:
```suggestion
    /// Convert a term to CNF
```

File: solver/src/models.rs [hunk 2]
```diff
@@ -1,6 +1,8 @@
 // Copyright 2022-2023 VMware, Inc.
 // SPDX-License-Identifier: BSD-2-Clause
 
+//! Parse the output of the SMT solver.
```
Comment (tchajed) [inline solver/src/models.rs:4]:
```suggestion
//! Parse and evaluate the models returned by the SMT solver.
```

File: solver/src/models.rs [hunk 3]
```diff
@@ -35,6 +37,7 @@ pub struct Model {
     pub symbols: HashMap<String, ModelSymbol>,
 }
 
+/// An evaluation error.
```
Comment (tchajed) [inline solver/src/models.rs:49]:
```suggestion
/// An error evaluating a function in an SMT solver model.
```

File: verify/src/safety.rs [hunk 4]
```diff
@@ -12,17 +12,25 @@ use fly::term::{fo::FirstOrder, prime::Next};
...
+    /// The invariant given in the module
     pub inv: Spanned<Term>,
+    /// The other invariants in the same proof as `inv`
     pub proof_invs: Vec<Spanned<Term>>,
 }
 
+/// An error that occured while constructing an invariant assertion
 #[derive(Error, Debug)]
 pub enum InvariantError {
+    /// Assertion was in some incorrect form
```
Comment (tchajed) [inline verify/src/safety.rs:30]:
```suggestion
    /// Assertion is not a safety property
```

**Review by tchajed — COMMENTED** [review]
I didn't quite review everything but I did cover a lot of the documentation changes.

File: Cargo.toml [hunk 5]
```diff
@@ -1,47 +1,26 @@
...
+[workspace]
+members = [
+	"fly",
+	"bounded",
+	"smtlib",
+	"solver",
+	"inference",
+	"verify",
+	"temporal-verifier",
+]
```
Comment (tchajed) [inline Cargo.toml:10]: It seems like `resolver ="2"` is needed to suppress this warning:
```
warning: some crates are on edition 2021 which defaults to `resolver = "2"`, but virtual workspaces default to `resolver = "1"`
note: to keep the current resolver, specify `workspace.resolver = "1"` in the workspace root's manifest
note: to use the edition 2021 resolver, specify `workspace.resolver = "2"` in the workspace root's manifest
```

File: Cargo.toml [hunk 6]
```diff
@@ -1,47 +1,26 @@
...
+
+[workspace.package]
 version = "0.1.0"
 edition = "2021"
 
-[features]
-trace = ["peg/trace"]
+# [features]
+# trace = ["peg/trace"]
```
[the remark this answers is not in the document]
Comment (Alex-Fischman) [reply Cargo.toml:17]: Is it in the fly crate currently, or in the root?

File: fly/src/parser.rs [hunk 7]
```diff
@@ -336,14 +336,12 @@ proof {
     }
 }
 
-#[cfg(test)]
-/// Parse a single term. Only used for testing.
+/// Parse a single term.
```
[the remark this answers is not in the document]
Comment (Alex-Fischman) [reply fly/src/parser.rs:339]: I was getting errors that the functions weren't visible in a test in a different crate. Removing cfg(test) fixed those errors.

[the remark this answers is not in the document]
Comment (tchajed) [reply fly/src/parser.rs:339]: Ok in that case I think it's fine to expose these two.

File: Cargo.toml (hunk 6 above)
[the remark this answers is not in the document]
Comment (tchajed) [reply Cargo.toml:17]: I don't understand your question - what I mean is those lines should be uncommented and moved to the fly crate's Cargo.toml file, to match the functionality they were providing before.

File: solver/src/models.rs [hunk 8]
```diff
@@ -16,24 +16,33 @@ use fly::{
...
 #[derive(Debug, Clone)]
 pub struct PartialInterp {
     universes: HashMap<String, Vec<String>>,
     // reverse of universes
     term_to_element: HashMap<String, Element>,
+    /// A map from relation names to a tuple of an interpretation and the return type
     pub interps: HashMap<String, (Interpretation, Sort)>,
 }
 
+/// Represents one relation in the parsed output of an SMT solver.
```
Comment (tchajed) [inline solver/src/models.rs:29]:
```suggestion
/// The value of a function in an SMT solver model.
```

File: solver/src/models.rs [hunk 9]
```diff
@@ -16,24 +16,33 @@ use fly::{
...
 pub struct ModelSymbol {
+    /// The arguments to the relation, and their sorts
     pub binders: Vec<(String, Sort)>,
+    /// The body of the relation that the solver solved
     pub body: Sexp,
+    /// The return sort of the relation
     pub ret_sort: Sort,
 }
 
+/// Represents the parsed Sat output of an SMT solver.
```
Comment (tchajed) [inline solver/src/models.rs:40]:
```suggestion
/// A parsed model from the SMT solver, without evaluation.
```

File: solver/src/models.rs [hunk 10]
```diff
@@ -16,24 +16,33 @@ use fly::{
...
     // reverse of universes
     term_to_element: HashMap<String, Element>,
+    /// A map from relation names to a tuple of an interpretation and the return type
     pub interps: HashMap<String, (Interpretation, Sort)>,
 }
 
+/// Represents one relation in the parsed output of an SMT solver.
 #[derive(Debug, Clone, Serialize, PartialEq, Eq)]
 pub struct ModelSymbol {
+    /// The arguments to the relation, and their sorts
```
Comment (tchajed) [inline solver/src/models.rs:32]:
```suggestion
    /// The bound arguments of the function, and their sorts.
```

File: solver/src/models.rs [hunk 11]
```diff
@@ -16,24 +16,33 @@ use fly::{
...
     pub binders: Vec<(String, Sort)>,
+    /// The body of the relation that the solver solved
```
Comment (tchajed) [inline solver/src/models.rs:34]:
```suggestion
    /// The body of the function
```

File: solver/src/models.rs [hunk 12]
```diff
@@ -16,24 +16,33 @@ use fly::{
...
     pub body: Sexp,
+    /// The return sort of the relation
```
Comment (tchajed) [inline solver/src/models.rs:36]:
```suggestion
    /// The return sort of the function
```

**Review by tchajed — COMMENTED** [review]
Here are a few more documentation fixes

File: Cargo.toml (hunk 5 above)
Comment (Alex-Fischman) [reply Cargo.toml:10]: Where is that warning? I'm not seeing it...

**Comment by Alex-Fischman** [comment]
> Could you stop adding any documentation in this PR, and turn off `#[warn(missing_docs)]` for now? It makes the PR much harder to review.

Sure, I'm just going to commit my documentation for MARCO, since I wrote the file. Then I'll comment out the #[warn] and open an issue to add it back in.

File: Cargo.toml (hunk 5 above)
Comment (Alex-Fischman) [reply Cargo.toml:10]: I'm still not seeing the warning, but https://doc.rust-lang.org/cargo/reference/resolver.html#resolver-versions makes it seem like it'll be good style anyway, so I'll add it.

**Pull request merged**